### PR TITLE
Commcare: replace superagent with common http helpers

### LINKED ISSUE #414: Commcare: don't use language-http
commcare currently depends on language http. It also uses common.http, which is a bit mad

We should refactor so that it uses the new http helpers in common

### PR DESCRIPTION
**Pull request by hunterachieng**
## Summary

Switch `commcare` adaptor implementation to use `fetch` for all http calls

## Implementation Details

Currently the `commcare` module utilizes superagent and `language-http` to handle its requests.  Our aim was to switch the implementation with normal fetch.

## How to Test

- [x] Uses the standard format for http requests defined in https://github.com/OpenFn/adaptors/blob/main/packages/template/src/Utils.js
- [x] Have no reference to  superagent or `language-http` . 

## Issues

Fixes #414

## Review Checklist

Before merging, the reviewer should check the following items:

- [ ] Does the PR do what it claims to do?
- [ ] If this is a new adaptor, has the migration tool been run and the
      migration guide followed?
- [ ] Are there any unit tests? Should there be?
- [ ] Is there a changeset associated with this PR? Should there be? Note that
      dev only changes don't need a changeset.

### CODE REVIEW

File: packages/commcare/src/Utils.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  const { body, ...responseWithoutBody } = response;
+  const nextState = {
+    ...composeNextState(state, response.body),
+    response: responseWithoutBody,
+  };
+
+  return callback(nextState);
+};
+
+export function request(state, method, path, data, params, header, authType) {
```
Comment (josephjclark) [inline packages/commcare/src/Utils.js:18]: This signature is getting very long! Instead of declaring all these parameters, I would just declare one: an object.

Basically just wrap the parameters up into an object, like this:
```
export function request({ state, method, path, data, params, header, authType })
```

File: packages/commcare/src/Adaptor.js [hunk 2]
```diff
@@ -203,29 +172,41 @@ export function submit(formData) {
...
+
+    console.log(
+      'getting from url: '.concat(
+        state.configuration.hostUrl,
+        `/a/${state.configuration.applicationName}/api/v0.5/configurablereportdata/${reportId}/`
+      )
+    );
+    console.log('with params: '.concat(params));
+
+    return await request({
```
Comment (josephjclark) [inline packages/commcare/src/Adaptor.js:186]: You should not mix `await` and `then`. Either use a promise chain (`then().then()`), or async/await.

Both styles are acceptable, just not at the same time :sweat_smile:

In this case you can just drop the `await` keyword and return the promise chain. That works nicely.

File: packages/commcare/src/Adaptor.js [hunk 3]
```diff
@@ -44,23 +33,21 @@ export function execute(...operations) {
...
+    state,
+    method: 'POST',
+    path: path,
+    data: body,
+    authType: 'basic',
+    contentType: 'text/html',
+    parseAs: 'text',
   });
+
+  return prepareNextState(state, response, s => s);
```
[the remark this answers is not in the document]
Comment (josephjclark) [reply packages/commcare/src/Adaptor.js:50]: Ah, to make this work more cleanly, you need to default the parameter value _inside_ `prepareNextState`

```
export const prepareNextState = (state, response, callback = s => s) =>  { ...}
```

File: packages/commcare/src/Utils.js [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+  const options = {
+    body: data,
+    headers: headers,
+    query: params,
+    parseAs: parseAs,
+  };
+
+  const url = `${hostUrl}${path}`;
+
+  return commonRequest(method, url, options).then(response => {
```
Comment (josephjclark) [inline packages/commcare/src/Utils.js:50]: I think you can just do:

```
return commonRequest(method, url, options).then(logResponse);
```

**Review by josephjclark — COMMENTED** [review]
A few comments to look into tomorrow Hunter!

It's generally a nice, clean, minimal PR. I think you can make it a bit thinner, and you've hit upon a couple of quirks of the old adaptor which we might need to address. I'll sleep on it and let you know tomorrow.

File: packages/commcare/src/Adaptor.js [hunk 5]
```diff
@@ -113,21 +98,20 @@ export function submitXls(formData, params) {
...
-      });
+
+    const response = await request({
+      state,
+      method: 'POST',
+      path: path,
+      data,
+      header: {
+        ...data.getHeaders(),
+        Authorization: `ApiKey ${username}:${apiKey}`,
```
[the remark this answers is not in the document]
Comment (hunterachieng) [reply packages/commcare/src/Adaptor.js:109]: Yes. I had a conversation with @mtuchi  but if I change it to use basic auth it might create disruptions in jobs using the adaptor so I think creating another issue to handle this would be ideal

File: packages/commcare/src/Adaptor.js [hunk 6]
```diff
@@ -113,21 +98,20 @@ export function submitXls(formData, params) {
...
+      },
+      authType: 'APIKEY',
```
[the remark this answers is not in the document]
Comment (hunterachieng) [reply packages/commcare/src/Adaptor.js:111]: I wrote that to distinguish the auth type since we have two different ways of doing auth. Let me refactor it

File: packages/commcare/src/Adaptor.js [hunk 7]
```diff
@@ -44,23 +33,21 @@ export function execute(...operations) {
...
-      .auth(username, password)
-      .set('Content-Type', 'application/xml')
-      .send(body)
-      .end((error, res) => {
-        if (!!error || !res.ok) {
-          reject(error);
-        }
-        resolve(res);
-      });
+async function clientPost({ state, path, body }) {
```
[the remark this answers is not in the document]
Comment (josephjclark) [reply packages/commcare/src/Adaptor.js:39]: I just wanna run this past @mtuchi, but the decision is:

- Let's break the API and move this function into utils. Call it `post`, which is a thin wrapper around `request`. It should not take state. Oh and do mark it `@private`
- Let's release this and #528 at the same time in the same major version

File: packages/commcare/src/Adaptor.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (josephjclark) [reply packages/commcare/src/Adaptor.js:105]: Great - raised as #554

File: packages/commcare/src/Adaptor.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (josephjclark) [reply packages/commcare/src/Adaptor.js:107]: Ok, nice idea, but I think right now it adds more confusion than clarity.

I've raised an issue to refactor auth - let's  make sure we have a nice solution there. In the meantime I recommend removing this line.

File: packages/commcare/src/Adaptor.js [hunk 8]
```diff
@@ -151,42 +131,23 @@ export function submitXls(formData, params) {
  * @returns {Operation}
  */
 export function submit(formData) {
-  return state => {
-    const jsonBody = expandReferences(formData)(state);
+  return async state => {
```
Comment (josephjclark) [inline packages/commcare/src/Adaptor.js:134]: I'm being picky but this doesn't need to be async anymore :)

File: packages/commcare/src/Adaptor.js [hunk 9]
```diff
@@ -203,29 +164,35 @@ export function submit(formData) {
...
-      console.log('with params: '.concat(params));
-      return params;
-    },
-    callback: function (state) {
-      var reportData = state.response.body;
-      return post(postUrl, {
-        body: reportData,
-      })(state).then(function (state) {
-        delete state.response;
+  return async state => {
```
Comment (josephjclark) [inline packages/commcare/src/Adaptor.js:167]: This one also shouldn't be async. We only use async if there's an `await` statement.

File: packages/commcare/src/Utils.js [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+  method,
+  path,
+  data,
+  params = {},
+  header = {},
+  authType,
+  contentType = 'application/json',
+  parseAs = 'json',
+}) {
+  const { hostUrl, username, password } = state.configuration;
```
Comment (josephjclark) [inline packages/commcare/src/Utils.js:29]: Ah this is why we need to pass state. hmm.

Ok, maybe we'll address this in the auth issue #554. Ignore my comments on state for now

Comment (josephjclark) [reply packages/commcare/src/Utils.js:29]: Basically I think we should pass an `auth` object, which is basically state.config, and we decide what auth to used based on the keys.

File: packages/commcare/src/Adaptor.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (josephjclark) [reply packages/commcare/src/Adaptor.js:39]: Hey @hunterachieng , sorry I should have posted in Friday - we're going to go ahead and do this! Please move and rename this function

File: packages/commcare/src/Adaptor.js [hunk 11]
```diff
@@ -44,23 +33,20 @@ export function execute(...operations) {
...
-        }
-        resolve(res);
-      });
+async function clientPost({ state, path, body }) {
+  const response = await request({
+    state,
+    method: 'POST',
+    path: path,
+    data: body,
+    contentType: 'text/html',
```
Comment (josephjclark) [inline packages/commcare/src/Adaptor.js:45]: Oh hey, this changes from `text/xml` to `text/html`. Is that correct?

File: packages/commcare/src/Adaptor.js [hunk 12]
```diff
@@ -203,29 +162,35 @@ export function submit(formData) {
...
-      return post(postUrl, {
-        body: reportData,
-      })(state).then(function (state) {
-        delete state.response;
+  return state => {
+    const path = `/a/${state.configuration.applicationName}/api/v0.5/configurablereportdata/${reportId}/`;
+
+    console.log('with params: '.concat(JSON.stringify(params)));
+
+    return request({
```
Comment (josephjclark) [inline packages/commcare/src/Adaptor.js:170]: I've just had an idea!

The nested promises here have always seemed weird to me. Nothing wrong with it, just wierd and a little hard to follow the logic.

Does it get any clearer if we re-write this whole chain with async/await?

**Review by josephjclark — COMMENTED** [review]
Hey @hunterachieng we're nearly there :)

Two extra things on top of the comments:

1) do you have a test job you're using against commcare? Could you check it in under `tests/integration/whatever.js`? This isn't a unit test, it's something that can be manually run with the cli. 

2) we need to generate a change set! There's notes in the readme but run `pnpm changeset` from root and add a major change with a note about what you've done, then check the file in.

File: packages/commcare/src/Adaptor.js (hunk 12 above)
Comment (josephjclark) [reply packages/commcare/src/Adaptor.js:170]: This is untested bit I was thinking about this:
```
const { body: reportData } = await request({
  state,
  method: 'GET',
  path,
  authType: 'basic',
})
const result = await request({
  state,
  method: 'POST',
  path: postUrl,
  params,
  data: reportData,
  authType: 'basic',
});

delete result.response;
console.log('fetchReportData succeeded.');
console.log('Posted to: '.concat(postUrl));
return prepareNextState(state, {});
```

File: packages/commcare/src/Utils.js [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+
+/**
+ * @private
+ *  * @function
+ * @param {string} path - Path to resource
+ * @param {object} data - Object which defines data that will be used to create a given instance of resource
+ * @param {string} contentType- Optional field for the header content-type
+ * @param {string} parseAs- Optional for data response parse i.e json
+ */
+export const post = async ({
```
Comment (josephjclark) [inline packages/commcare/src/Utils.js:16]: Ok, this isn't quite what I was expecting!

The `post` function will never work, because it doesn't have access to auth credentials.

I note that you've refactored Adaptor.js to not use `post` at all, but instead use request. This is fine but not what I expected.

Let's just remove the `post` function entirely, since it's broken and we're not using it :)

Comment (hunterachieng) [reply packages/commcare/src/Utils.js:16]: I tried using the post but the issue was that it has its own state function inside it and the function inside the adaptor. When I run it, I get undefined
